### PR TITLE
KAFKA-9297; CreateTopic API do not work with older version of the request/response

### PR DESCRIPTION
**Pull request by dajac**
The create topic api do not work with older version of the api. It can be reproduced by trying to create a topic with `kafka-topics.sh` from 2.3. It timeouts.

https://github.com/apache/kafka/commit/b94c7f479b917d4ec602c31a24f11390627c479b has added a check which raises an exception if a field has been set to a non-default value unless the field is marked as "ignorable".

The fields added in the version 5 of the response are always set regardless of the version used by the client. If an older version is used, an exception is thrown during the serialization because the fields have non-default values. We should either not set the fields for older versions in the api layer or mark them as ignorable. I have chosen the later in this case because it looks cleaner.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dajac** [comment]
retest this please

**Comment by dajac** [comment]
retest this please

**Comment by dajac** [comment]
retest this please

**Comment by soondenana** [comment]
Hi David, 

Thanks for fixing it. Some of these fields are not new. Maybe the description of PR need to change to reflect what exactly caused this issue (which is addition of check in Jason's PR).

File: core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala [hunk 1]
```diff
@@ -133,4 +140,35 @@ class CreateTopicsRequestTest extends AbstractCreateTopicsRequestTest {
     val response = sendCreateTopicRequest(req, notControllerSocketServer)
     assertEquals(1, response.errorCounts().get(Errors.NOT_CONTROLLER))
   }
+
+  @Test
+  def testOldestCreateTopicsRequest(): Unit = {
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala]: Should we name this `testAllVersionsCreateTopicsRequest()` or `testCreateTopicsRequestVersions()`?

File: core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala [hunk 2]
```diff
@@ -133,4 +140,35 @@ class CreateTopicsRequestTest extends AbstractCreateTopicsRequestTest {
...
+      ).asJava.iterator()))
+
+      val request = new CreateTopicsRequest.Builder(data).build(version.asInstanceOf[Short])
+      val response = sendCreateTopicRequest(request)
+
+      val topicResponse = response.data.topics.find(topic)
+      assertNotNull(topicResponse)
+      assertEquals(topic, topicResponse.name())
+      assertEquals(Errors.NONE.code(), topicResponse.errorCode())
+      if (version >= 5) {
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala:162]: Should we also set some topic configs?

**Review by mimaison — COMMENTED** [review]
Good catch! I've left a couple of some minor comments.

**Comment by ijuma** [comment]
Thanks for the PR. Can we add a unit test in the clients module too?

File: core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala (hunk 1 above)
Comment (dajac) [reply core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala]: `testCreateTopicsRequestVersions` sounds good to me.

File: core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala (hunk 2 above)
Comment (dajac) [reply core/src/test/scala/unit/kafka/server/CreateTopicsRequestTest.scala:162]: Yes. I have added it.

**Comment by dajac** [comment]
> Thanks for fixing it. Some of these fields are not new. Maybe the description of PR need to change to reflect what exactly caused this issue (which is addition of check in Jason's PR).

@soondenana You are absolutely right. I have updated the description to be more accurate.

**Comment by dajac** [comment]
@ijuma Sure. I have extended the tests in RequestResponseTest to cover all the versions.

**Comment by dajac** [comment]
@mimaison Thanks for the review. I have addressed your comments.

**Review by hachikuji — APPROVED** [review]
Agreed it makes sense to make these fields ignorable. The PR LGTM. Will let @mimaison do a final review and merge.

**Review by mimaison — APPROVED** [review]

**Pull request merged**